File: src/lucin/telemetry.py
"""Lucin telemetry — anonymous, aggregate-only, on by default.

WHAT IS SENT (allowlisted, enforced again server-side — see telemetry-worker/src/index.js):
  anon_id, event_type, lucin_version, python_version, os, frameworks,
  agent_count, tool_count, file_count, scan_duration_ms, output_format,
  ci_mode, finding_counts_json ({rule_id: count} — rule IDs and counts only),
  error_type.

WHAT IS NEVER SENT: file paths, target/repo names, source code, secret values,
witness text, tool or agent names. A security scanner whose own telemetry could
leak the secrets it finds would be the exact "lethal trifecta" pattern this
product exists to catch — the allowlist in the worker enforces this even if
this module is ever changed to send more.

Default ON, disclosed on first run, with an explicit off-switch:
  LUCIN_TELEMETRY=0 environment variable, or `lucin scan --no-telemetry`.
Config (anon id + opt-out + whether the disclosure banner has been shown) is
persisted at ~/.lucin/config.json, the same directory pinning.py already uses.
"""

import json
import os
import platform
import sys
import uuid
from pathlib import Path
from urllib import request as _urlrequest
from urllib.error import URLError

CONFIG_DIR = Path.home() / ".lucin"
CONFIG_PATH = CONFIG_DIR / "config.json"
COLLECTOR_URL = "https://lucin-telemetry.candura-telemetry.workers.dev/v1/event"
TIMEOUT_SECONDS = 1.5


def _load_config() -> dict:
    if CONFIG_PATH.exists():
        try:
            return json.loads(CONFIG_PATH.read_text())
        except (json.JSONDecodeError, OSError):
            pass
    return {}


def _save_config(cfg: dict) -> None:
    try:
        CONFIG_DIR.mkdir(parents=True, exist_ok=True)
        CONFIG_PATH.write_text(json.dumps(cfg, indent=2))
    except OSError:
        pass  # never let telemetry bookkeeping break the CLI


def _anon_id(cfg: dict) -> str:
    if "anon_id" not in cfg:
        cfg["anon_id"] = uuid.uuid4().hex
        _save_config(cfg)
    return cfg["anon_id"]


def is_enabled() -> bool:
    if os.environ.get("LUCIN_TELEMETRY") == "0":
        return False
    cfg = _load_config()
    return cfg.get("enabled", True)


def disable() -> None:
    cfg = _load_config()
    cfg["enabled"] = False
    _save_config(cfg)


def enable() -> None:
    cfg = _load_config()
    cfg["enabled"] = True
    _save_config(cfg)


def maybe_print_first_run_notice(console) -> None:
    """Print the disclosure exactly once, on the first run ever. Never blocks."""
    cfg = _load_config()
    if cfg.get("notice_shown"):
        return
    cfg["notice_shown"] = True
    _anon_id(cfg)
    _save_config(cfg)
    if not is_enabled():
        return
    console.print(
        "[dim]Lucin sends anonymous usage stats (version, OS, which rules fire, "
        "counts only — never file paths, code, or finding content) to help "
        "prioritize development. Disable with `--no-telemetry`, "
        "`LUCIN_TELEMETRY=0`, or `lucin telemetry disable`. "
        "Details: `lucin telemetry status`.[/dim]"
    )
    console.print()


def build_scan_event(result, output_format: str, ci: bool) -> dict:
    frameworks = sorted({a.framework for a in result.agents if getattr(a, "framework", None)})
    finding_counts: dict[str, int] = {}
    for f in result.findings:
        finding_counts[f.id] = finding_counts.get(f.id, 0) + 1
    return {
        "event_type": "scan",
        "lucin_version": _lucin_version(),
        "python_version": f"{sys.version_info.major}.{sys.version_info.minor}",
        "os": platform.system().lower(),
        "frameworks": ",".join(frameworks)[:64],
        "agent_count": len(result.agents),
        "tool_count": sum(len(a.tools) for a in result.agents),
        "file_count": len({f.source_file for f in result.findings if f.source_file}),
        "scan_duration_ms": result.scan_duration_ms,
        "output_format": output_format,
        "ci_mode": 1 if ci else 0,
        "finding_counts_json": finding_counts,
    }


def build_command_event(command: str) -> dict:
    """A minimal event for commands other than `scan` — just proves the command ran.

    No arguments, paths, or discovered content are ever included here; that's the
    whole reason `discover` (which enumerates MCP configs across every IDE on the
    machine) is safe to instrument the same way as everything else.
    """
    return {
        "event_type": f"cmd_{command}",
        "lucin_version": _lucin_version(),
        "python_version": f"{sys.version_info.major}.{sys.version_info.minor}",
        "os": platform.system().lower(),
    }


def build_error_event(exc: BaseException) -> dict:
    return {
        "event_type": "error",
        "lucin_version": _lucin_version(),
        "python_version": f"{sys.version_info.major}.{sys.version_info.minor}",
        "os": platform.system().lower(),
        "error_type": type(exc).__name__,
    }


def _lucin_version() -> str:
    from lucin import __version__
    return __version__


def last_event() -> dict | None:
    """The exact payload built for the most recent command — what `lucin telemetry
    status` shows. Recorded locally even when telemetry is disabled or the send
    fails, so a user can always inspect what *would* be sent."""
    return _load_config().get("last_event")


def send_event(event: dict) -> None:
    """Fire-and-forget. Never raises, never blocks the CLI beyond TIMEOUT_SECONDS."""
    cfg = _load_config()
    event = dict(event)
    event["anon_id"] = _anon_id(cfg)
    cfg["last_event"] = event
    _save_config(cfg)
    if not is_enabled():
        return
    try:
        body = json.dumps(event).encode("utf-8")
        req = _urlrequest.Request(
            COLLECTOR_URL, data=body,
            # Cloudflare returns 403 to the default "Python-urllib/x.y" UA (looks
            # like a generic bot signature) — a real UA is required for delivery,
            # not just politeness. Found by testing: silently swallowed by the
            # except clause below until this was diagnosed with a direct request.
            headers={"Content-Type": "application/json", "User-Agent": f"lucin-cli/{_lucin_version()}"},
            method="POST",
        )
        _urlrequest.urlopen(req, timeout=TIMEOUT_SECONDS).close()
    except (URLError, OSError, ValueError):
        pass  # telemetry must never break or slow down a scan meaningfully

File: site/functions/api/telemetry.js
// Lucin telemetry collector, as a Cloudflare Pages Function.
//
// Strict allowlist by design: this endpoint only accepts the fields listed in
// ALLOWED_STRING/ALLOWED_NUMBER below and silently drops everything else. That
// is the enforcement point for "never transmits file paths, source code,
// secret values, or tool/agent names" — even if the CLI is ever changed
// (by mistake or otherwise) to send more, this endpoint cannot persist it.
//
// No IP address, user agent, or any other request metadata is stored.
// Lives at lucin.pages.dev/api/telemetry — same origin as the site, so no
// separate subdomain (and no leaked project name) is needed.

const ALLOWED_STRING = new Set([
  "anon_id", "event_type", "lucin_version", "python_version", "os",
  "frameworks", "output_format", "error_type",
]);
const ALLOWED_NUMBER = new Set([
  "agent_count", "tool_count", "file_count", "scan_duration_ms", "ci_mode",
]);
const RULE_ID_RE = /^AG-[A-Z0-9-]+$/;
const MAX_STRING_LEN = 64;

function sanitizeEvent(body) {
  const out = {};
  for (const key of ALLOWED_STRING) {
    const v = body[key];
    if (typeof v === "string" && v.length > 0 && v.length <= MAX_STRING_LEN) {
      out[key] = v;
    }
  }
  for (const key of ALLOWED_NUMBER) {
    const v = body[key];
    if (typeof v === "number" && Number.isFinite(v)) {
      out[key] = v;
    }
  }
  if (body.finding_counts_json && typeof body.finding_counts_json === "object") {
    const counts = {};
    for (const [ruleId, count] of Object.entries(body.finding_counts_json)) {
      if (RULE_ID_RE.test(ruleId) && typeof count === "number" && count >= 0 && count < 100000) {
        counts[ruleId] = Math.floor(count);
      }
    }
    out.finding_counts_json = JSON.stringify(counts);
  }
  return out;
}

export async function onRequestPost({ request, env }) {
  let body;
  try {
    body = await request.json();
  } catch {
    return new Response("bad json", { status: 400 });
  }

  if (typeof body !== "object" || body === null || typeof body.anon_id !== "string" || typeof body.event_type !== "string") {
    return new Response("missing anon_id/event_type", { status: 400 });
  }

  const clean = sanitizeEvent(body);

  const columns = Object.keys(clean);
  const placeholders = columns.map(() => "?").join(", ");
  const sql = `INSERT INTO events (${columns.join(", ")}) VALUES (${placeholders})`;

  try {
    await env.DB_TELEMETRY.prepare(sql).bind(...columns.map((c) => clean[c])).run();
  } catch (e) {
    return new Response("db error", { status: 500 });
  }

  forward(env, clean);
  return new Response(null, { status: 204 });
}

// Mirror CLI events into the same vendor as site/functions/api/events.js, so
// scans and site behaviour are queryable side by side rather than in two tools.
// Only the already-sanitised `clean` object is forwarded — the allowlist above
// is still the only thing that decides what may leave this Worker.
// Fire-and-forget: a vendor outage must not fail a write that already succeeded.
function forward(env, clean) {
  if (!env.POSTHOG_KEY) return;
  const host = env.POSTHOG_HOST || "https://eu.i.posthog.com";
  fetch(`${host}/capture/`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      api_key: env.POSTHOG_KEY,
      event: `cli_${clean.event_type || "unknown"}`,
      distinct_id: clean.anon_id || "anon",
      properties: { ...clean, source: "cli" },
    }),
  }).catch(() => {});
}
